azure/ConsiderationsForServiceDesign.md

```diff
@@ -0,0 +1,99 @@
+
+### Use Previews to Iterate 
+ Before releasing your API plan to invest significant design effort, get customer feedback, & iterate through multiple preview releases. This is especially important for V1 as it establishes the abstractions and patterns that developers will use to interact with your service.
+
+> :ballot_box_with_check: **YOU SHOULD**  write and test hypotheses about how your customers will use the API.
+> 
+> :ballot_box_with_check: **YOU SHOULD**  release and evaluate a minimum of 2 preview versions prior to the first GA release.
+> 
+> :ballot_box_with_check: **YOU SHOULD**  identify key scenarios or design decisions in your API that you want to test with customers, and ask customers for feedback and to share relevant code samples.
+> 
+> :ballot_box_with_check: **YOU SHOULD**  consider doing a *code with* exercise in which you actively develop with the customer, observing and learning from their API usage.
+> 
+> :ballot_box_with_check: **YOU SHOULD**  capture what you have learned during the preview stage and share these findings with your team and with the API Stewardship Board.
+
+### Avoid Surprises
+A major inhibitor to adoption and usage is when an API behaves in an unexpected way. Often, these are subtle design decisions that seem benign at the time, but end up introducing significant downstream friction for developers.
+
+> :ballot_box_with_check: **YOU SHOULD** avoid polymorphism, especially in the response. An endpoint __SHOULD__ work with a single type to avoid problems during SDK creation.
+>
+> :ballot_box_with_check: **YOU SHOULD** make [Collections](./Guidelines.md#collections) easy to work with. Collections are a common source of review comments. It is important to handle them in a consistent manner within your service.
+>
+> :ballot_box_with_check: **YOU SHOULD** return a homogeneous collection (single type).  Do not return heterogeneous collections unless there is a really good reason to do so.  If you feel heterogeneous collections are required, discuss the requirement with an API reviewer prior to implementation.
+>
+> :ballot_box_with_check: **YOU SHOULD** support server-side paging, even if your resource does not currently need paging. This avoids a breaking change when your service expands.
+
+### Design for Change Resiliency 
+As you build out your service and API, there are a number of decisions that can be made up front that add resiliency to client implementations. Addressing these as early as possible will help you iterate faster and avoid breaking changes.
+
+> :ballot_box_with_check: **YOU SHOULD** use extensible enumerations. Extensible enumerations are modeled as strings - expanding an extensible enumeration is not a breaking change. 
```

Comment:
I think this is oversimplifying the issue. Expanding a non-extensible enum can also be non-breaking and expanding extensible enum can be breaking (e.g. if the enum is used as a discriminator for a polymorphic hierarchy).

Reply:
It's true that there are a number of subtleties that constitute a breaking change. I don't think we should attempt to tackle them all here, but then should provide strong guidance that covers the majority of the use cases.

Reply:
@KrzysztofCwalina This is interesting!  Since changing a discriminator value is a "hard breaking change" (not even allowed with API version change except with special approval), it is an error to define a discriminator field as an extensible enum.  This should be caught by our linters -- but I'm guessing that it isn't.

Reply:
We have to allow the discriminator to be an extensible enum because this is actually a common case. depending on the usage, this may or may not be a problem for old clients. Client SDKs have to be resilient against this and the pattern I describe in my video is the way to make this work: https://youtu.be/ycoXRuTdfNo?list=PL9XzOCngAkqs4m0XdULJu_78nM3Ok3Q65&t=771

An old client will either have to ignore an instance of an unknown discriminator type or the CUSTOMER (not SDK) can use an old SDK but look for the new discriminator value and manually parse/format the JSON to work with it. It's not a great story but I see no way to make the story better.